File: src/algorithms/binary-trees/invert-binary-tree/solution2.js
// Recursive Solution
// O(n) time | O(d) space
function invertBinaryTree(tree) {
  if (tree === null) return;
  swapLeftAndRight(tree);
  invertBinaryTree(tree.left);
  invertBinaryTree(tree.right);
}

function swapLeftAndRight(tree) {
	const left = tree.left;
	tree.left = tree.right;
	tree.right = left;
}

exports.invertBinaryTree = invertBinaryTree;

File: src/algorithms/binary-trees/invert-binary-tree/solution2.go
package main

type BinaryTree struct {
	Value int

	Left *BinaryTree
	Right *BinaryTree
}

// O(n) time | O(d) space
func (tree *BinaryTree) InvertBinaryTree() {
	tree.Left, tree.Right = tree.Right, tree.Left
	if tree.Left != nil {
		tree.Left.InvertBinaryTree()
	}
	if tree.Right != nil {
		tree.Right.InvertBinaryTree()
	}
}